### ISSUE TITLE
Error messages, I18N and Namespaces

### ISSUE THREAD

**Issue by apotonick** [hanami/validations#67]
Reform 2.0 does support lotus-validations and I will push towards making this the default implementation. Thanks for this. :heart: 

As discussed with Luca, I'm currently designing the process of generating an (i18nized) message string from an `Error` object. I do this in Reform first, and will then extract it to this gem.

Consider the following object for a nested field `band.label.name`. 

```ruby
#<Lotus::Validations::Error:0x007fb6d5b48aa0 @attribute_name="name", 
  @validation=:presence, @expected=true, @actual=nil, 
  @namespace="band.label", @attribute="name">]}
```

In order to translate this, I propose the following I18N paths (please comment!).

```
form.band.label.name.blank
form.band.name.blank
form.name.blank
```

Don't worry about the prefix `form.`, this could be anything and is just an example.

Another thing is: How to name the suffixes, e.g. `blank`? This has to be computed from `Error` by looking at `@validation` and `@actual`, so _we need to specify translation suffixes_ for validation/result tuples.

Any input?

**Comment by apotonick** [comment]
So, to make it a bit clearer: The API I'm dreaming of is as follows.

```ruby
error #=> <Lotus::Validations::Error @attribute_name="name", ..., @namespace="band.label">

error.to_s 
# invokes
error.to_i18n_s # or something like that
# will compute the aforementioned i18n paths, find the string, etc. and RETURN AN ERROR MESSAGE. 
```

My point here is that Reform will be responsible to provide the `@namespace`, then calls `Error#to_s` and the rest happens in Lotus. Is that cool?

**Comment by jodosha** [comment]
@apotonick Thanks for following up our discussion.

My idea for Lotus::Validations as a standalone lib, was something like this:

```ruby
if defined?(I18n)
  def to_s
    I18n.translate(i18n_key, message)
  end
else
  def to_s
    message
  end
end

private

def i18n_key
  # ...
end

def message
  # ...
end
```

If `I18n` gem is present, try to find a translation for the given `i18n_key` (eg. `"form.name.blank"`), otherwise fallback to `message`. This last method is a default error message, such as `"Name can't be blank"`.

I'd get rid of `#to_i18n_s`. Imagine you're iterating a collection of errors how to decide which method to use? By exposing only `#to_s` has the advantage of doing this error presentation always using the same method.

By overriding `#to_s` we can leverage the fact that template engines will call that method implicitly.

```erb
<% form.errors.each do |error| %>
  <%= error %> <!-- implicit #to_s call -->
<% end %>
```

To recap, L::V is able to compute default english error messages (`#message`), devs can customize a message via I18n or fallback to the standard message. All this logic is exposed via `#to_s`.

Sounds good? @apotonick @lotus/core-team

**Comment by apotonick** [comment]
Oh yeah absolutely, @jodosha ! I as a Trb representant am only interested in the `#to_s` method, so I'm totally fine with a conditional `to_s` as long as I get a message back! :wink:

The `to_i18n_s` was more a conceptual thing...

Anyways, what about the namespace? Would that work for you the way I specified it? I don't really care about the details which paths get searched when, etc, but if you're fine I will implement a prototype and then we'll see? :sparkles:

**Comment by jodosha** [comment]
@apotonick 

#### Namespace

For the namespace there is a pending issues about [composable validations](https://github.com/lotus/validations#composable-validations).

If we have a module `EmailValidations` that is included into `Signup`. Which is the expected starting value for the namespace: `email_validations.*` or `signup.*`? What if I want to have two different messages for these two cases? If we exclude the class/module name from the first value of the namespace, we can do something clever with [I18n scopes](https://github.com/svenfuchs/i18n/blob/master/lib/i18n.rb#L51-L72).

```ruby
I18n.translate(i18n_key, scope: namespaces, default: message)

puts i18n_key       # => "email.presence"
puts namespaces # => ["signup", "email_validations"]
puts message       # => "Email can't be blank"
```

In this way if a translation for `signup.email.presence` is defined it will return it, otherwise it will try with `email_validations.email.presence` or eventually fall back to the default message.

#### Validation names

Regarding the validation names, I'd stick with the ones who are already implemented. For instance, instead of using `blank`, we should use `presence`. In this way developers can have a symmetry between the names that they use to define the validations and their counterpart error messages.

Eg.

```ruby
class Signup
  include Lotus::Validations

  attribute :email, presence: true
end
```

It will correspond to the following custom message:

```yaml
en:
  signup:
    email:
      presence: "Email can't be blank"
```

What do you think?

**Comment by apotonick** [comment]
Reverse order: 

### Validation names

The less mapping, the better! Happy to use `presence` instead of `blank`, etc. Let's see where we go, I will play with it tomorrow.

### Namespaces 

I do not really understand how composable validations are related to namespaces? :confetti_ball: The `:scope` option of I18N is perfect, thanks for pointing that out.

Let me play with it a bit, we're on the same page and once we got some examples we can figure out details, e.g. where do class names go, etc. Thanks :beers:

**Comment by jodosha** [comment]
> I do not really understand how composable validations are related to namespaces?

I assume that the top level entry for a namespace is the name of the class. Eg. if you have `Signup` that includes L::V, the namespace will start with `signup` and the complete name will be `signup.email.presence` etc..

Now, composable validations are a feature that encourages code reusability. As a dev, I want to have the freedom to define the message once, but be able to specify an exception. Let me put in code terms. If I have a module like this:

```ruby
module EmailValidations
  include Lotus::Validations

  attribute :email, presence: true
end
```

Then I define I18n message for it.

```yaml
en:
  email_validations:
    email:
      presence: "Please enter email"
  signup:
    email:
      presence: "Email address is missing"
    age:
      inclusion: "You must be at least 18"
```

So I can reuse it in a few places:

```ruby
class UpdateUserDetails
  include EmailValidations
end

class ForgotPassword
  include EmailValidations
end

validator = UpdateUserDetails.new(email: nil)
validator.valid? # => false
validator.errors.first.to_s # => "Please enter email"

# ..and

validator = ForgotPassword.new(email: '   ')
validator.valid? # => false
validator.errors.first.to_s # => "Please enter email"
```

Now I have a `Signup` validator, but I want to specify a different message, because the key `en.signup.email.presence` is defined, I expect it to be used.

```ruby
class Signup
  include Lotus::Validations
  include EmailValidations

  attribute :age, inclusion: 18..999
end

validator = Signup.new(email: nil, age: 1)
validator.valid?
validator.errors.first.to_s # => "Email address is missing"
validator.errors.last.to_s  # => "You must be at least 18"
```

If we're able to track both the module where it was defined (eg `EmailValidations`) and the current class (eg `Signup`) and use them as I18n scopes, we can make working the scenario described above.

**Comment by solnic** [comment]
Having error objects knowing how to dump themselves into a string message sounds cool but it's not gonna be so nice when you want a complex error message that needs more context than just the plain error object. I believe having a separate error obj => error msg mapping layer could be more powerful and simpler. Otherwise you're gonna add a lot of logic to errors making them too big and complex.

**Comment by jeremyf** [comment]
This is from a [localization and translation planning document](https://github.com/ndlib/sipity/blob/c78173fe00844b6f979415443564a8a50f1a283b/architecture/localization-and-translation.md) that I spent some time on:

# Localization and Translation

The processing actions are represented through a Form object.
Each form object is associated with a processing entity.
At present processing entities represent one of three things:

* Sipity::Models::Work
* Sipity::Models::WorkArea
* Sipity::Models::SubmissionWindow

Instead of managing all of localizations of the different form objects through separate translation keys,
I want to see these unified and predictable.

I am assessing the gnarly localization/translation needs:

* **Form Labels** for input fields on "command pages"
* **Form Hints** input fields on "command pages"
* **Show Labels** for predicates on "query pages"
* **Form Error Messages** on "command pages"

As I've looked through the implementation of ActiveModel::Name and SimpleForm
there are a few things:

## Form Labels

SimpleForm uses the :object_name derived from the `:as` option or `#param_key`.
This is known as the form's `object_name`.
When posting a form, the object_name is the hash key for the posted attributes.

- `simple_form_for :user` would be 'user'
- `simple_form_for @user` would be 'user' (assuming User.model_name.param_key == 'user')
- `simple_form_for @user, as: :ketchup` would be 'ketchup'

If there is no hit on the `object_name` and attribute name SimpleForm will try
the `defaults` and attribute name.
If that fails, SimpleForm will pass the responsibility up to the object class's `human_attibute_name` method.
(Yes, the Law of Dememeter is being violated with that concession).

## For Hints

As the Labels, but there is no fallback on `human_attribte_name` of the object's class.

## Show Labels

The logical method is to fallback on the `human_attribute_name` of the object's class.

## Form Error Messages

Leverages `errors` from the underlying model, which in turn goes through the `.lookup_ancestors` of the object's class.

# Plan

For each Form object, expose a `#model_name` object.
That `#model_name` object would return an `ActiveModel::Naming` instance that was built with an object that responds to:

* `#name` (and returns the concept it was using: Work, SubmissionWindow, WorkArea)
* `#lookup_ancestors` (and returns `[]`, is that correct?)
* `#i18n_scope` (and returns the correct scope)

This leverages the implementation details of `ActiveModel::Name`.
I am hesistant to choose this implementation, but I'd prefer this over implementing the ActiveModel::Name interface.
If I implemented the ActiveModel::Name interface I would miss out on the fallback of `human_attribute_name`; Which will
rescue me from insanity.

Going forward, I would want to push the :hint behavior to the form object:

```ruby
<%= f.input :title, hint: f.object.hint(:title) %>
```

**Comment by jeremyf** [comment]
And here is [the implementation](https://github.com/ndlib/sipity/blob/c78173fe00844b6f979415443564a8a50f1a283b/app/presenters/sipity/controllers/translation_assistant.rb).

I opted to take inspiration from RDF regarding nomenclature (http://www.w3.org/RDF/).

**Comment by apotonick** [comment]
@solnic Really good point! What about a very very very basic `#to_s` implementation? You can still use the `Error` instances outside if you want to compile complex messages that need multiple errors.

We do that in Reform all the time: run custom validations that access `errors` and then `add` new, "custom-made" errors.

@jeremyf I would refrain from using the ActiveModel API - I find it incredibly convoluted and complex, especially the `lookup_ancestors` is simply not the right place in a validator object. Is that the right thread for this to talk about that? :grin:

**Comment by jodosha** [comment]
@solnic @apotonick Which is the integration point for that? I remember you that the basic goal is to provide a simple usage like this:

```erb
<% validator.errors.each do |error| %>
  <%= error %>
<% end %>
```

This will give the power of get both standard or customized error messages.

Because `Error` has informations such as attribute name, nested namespace, violated validation, expected and current value. One solution could be that devs will _present_ errors according to their custom logic: eg `<%= print_error(validator, error) %>`.

I'm open to code suggestions for advanced cases. But again keep in mind that we aim for trivial usage for developers. :wink:

**Comment by solnic** [comment]
I wouldn't make error responsible for coercing itself to string and I don't really care it'd be "idiomatic ruby". Generating error messages can be pretty complex and I'd prefer to have a separate object responsible for that. I would also prefer to have one way of serializing errors into string messages vs "oh, if it's simple you just `to_s` otherwise you do something else" because it would result in more complexity (you would have to make error responsible for simple string coercion and build something else for complex cases too, pointless IMO).

Regarding "trivial usage" - look how ActiveRecord ended, it's been always optimized for being (seemingly) simple. Pretty DSLs on top of tons of complexity, as somebody tweeted recently, is not the way to go :)

As far as code suggestions goes, if we're talking about generating HTML with error messages I believe a simple decorator would be an elegant solution:

``` erb

# excuse that really horrible name, just showing the concept
<% ErrorHTMLPresenter.new(errors).each do |error| %>
  <% error %>
<% end %>
```

Notice that with a decorator approach you can add a lot of really nice and convenient functionality w/o cluttering the error object. IE you can configure error html templates, so that a list of messages is formatted properly etc. A lot can be done easily and error API would stay simple.

**Comment by apotonick** [comment]
I like your point about not polluting the `Error#to_s` method, @solnic.

What about this?

```ruby
class Error
  def to_s
    Renderer.new(self).()
  end
```

This delegation is minimalising the pollution of `Error` and, at the same time, allows other decorators to be used.

**Comment by solnic** [comment]
@apotonick this couples error to a specific renderer implementation :(

**Comment by apotonick** [comment]
@solnic Yeah but somewhere the messages need to be compiled _per default_, and as I said, you could do as follows if you don't like my specific renderer.

```ruby
Formtastic::Errors::Renderer.new(errors).()
```

**Comment by solnic** [comment]
@apotonick wdym that they need to be compiled per default? they need to be compiled whenever you need to display them, so I don't see any reason why you would want to couple `to_s` to a specific rendering implementation.

**Comment by apotonick** [comment]
Haha I love this discussion! :laughing:  Lotus _has_ to ship with one specific implementation, otherwise the validations are useless for most people - you can't tell people to write their own message compilation with I18N, etc. (is that what you mean or am I wrong here?).

I totally agree that this should _not_ sit directly in `Error#to_s` but in a separate renderer class. Now, all we need to do is connect errors with renderer, and I don't see why this is problematic to do so in `Error#to_s`, given that you could easily change the default renderer???

**Comment by solnic** [comment]
If you do `def to_s; Renderer.new; end` then it doesn't feel like it's easy to swap with a different strategy. I'm also not sure why lotus has to ship with one specific implementation. To me it seems like an integration thing - lotus-validations can be integrated with a view rendering layer or a json api layer etc.

**Comment by jodosha** [comment]
@solnic @apotonick I see in this case `#to_s` as a _textual, human friendly representation of the internal state of the error object_. For this I don't see nothing wrong to implement this strategy there.

Well, in an ideal world I would have chosen a presenter strategy like:

```ruby
ErrorPresenter.new(error).to_s
```

But, this isn't possible because there isn't a hook mechanism to do this internally. The only solution that I see at the moment is to use a syntax like this:

```ruby
<% validators.errors.each do |error| %>
  <%= ErrorPresenter.new(error) %> <--! #to_s is implicitly called by ERB -->
<% end %>
```

For me this syntax is a :-1: because too verbose. However, devs are free to implement their own presentation strategy, by wrapping the error and not relying on `#to_s`.

> If you do def to_s; Renderer.new; end then it doesn't feel like it's easy to swap with a different strategy. 

Because `Error` is an object that is instantiated internally to a validator, there isn't an easy way to apply Dependency Injection. One way could be to accept a class that represent the error type that we want to use. Eg `Signup.new(error: MyCustomErrorType)`. But this will make things complicated when the validator isn't explicitly instantiated by the dev. This is the case of Lotus::Action's params.

> I'm also not sure why lotus has to ship with one specific implementation.

Well, if we don't ship one specific implementation means two things:

  1. No error messages at all
  2. Support for an error message presenter interface to be implemented by developers.

Because this is a common problem, it's up to a framework to solve it (like @apotonick suggested).

@solnic I'm still struggling to understand, what's your idea for error messages, from a syntax perspective. Again, I value a lot your opinion, that's why I'd love to hear more from you. Thanks.

**Comment by solnic** [comment]
lotus-validations could ship with integration for various use cases like rendering error messages in html templates, json responses, using I18n etc. There's absolutely no need for `Error#to_s` to exist except for debugging inspection maybe.

You can have an "error rendering plugin" for validators that could do something like `validator.error_messages.each` but keep `errors` pure.

Also remember that errors can be nested, messages may require data from other nodes and in general this is a pretty complicated problem. Trying to convince yourself that it can be solved as a simple `to_s` as a default behavior "because it's simple" doesn't feel right :) For me it's a rabbit hole right there. Hence I'm suggesting an external API to handle that and using error objects as simple pure data/information providers without much logic.

**Comment by pascalbetz** [comment]
My two cents:

I don't think that ```to_s``` is the right way to get i18n aware error messages. Error objects should just contain the information that can be used to render a user friendly error message.
If the syntax for views is too verbose, then this can be solved by a helper.

**Comment by sockmonk** [comment]
So today, is there any way at all to specify a custom user-facing error message with lotus validations? I'd love to at least consider using it, but this would be a deal breaker if there isn't any way at all yet. At least until this gets solved.

**Comment by joneslee85** [comment]
@sockmonk not atm

**Comment by apotonick** [comment]
I started working on it yesterday, here is my fork branch: https://github.com/trailblazer/lotus-validations/tree/error-messages

**Comment by joneslee85** [comment]
@apotonick I want to :kiss: you

**Comment by joneslee85** [comment]
@apotonick can you please create a PR so we could follow your progress? TIA

**Comment by apotonick** [comment]
Once it's presentable! :kiss:

**Comment by fbernier** [comment]
Hey, anything we can do to help ship this?

**Comment by jodosha** [comment]
@fbernier I've started working on this. Please have a look at: https://github.com/lotus/validations/tree/error-messages

**Comment by jodosha** [comment]
Closed by #100
